pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1218,7 +1223,22 @@ public void run(Timeout timeout) throws Exception {
     };
 
     @Override
-    protected void flush() {
+    public CompletableFuture<Void> flushAsync() {
+        if (isBatchMessagingEnabled()) {
+            CompletableFuture<MessageId> lastSendFuture;
+            synchronized (ProducerImpl.this) {
+                batchMessageAndSend();
+                lastSendFuture = this.lastSendFuture;
+            }
+            if (null != lastSendFuture) {
+                return lastSendFuture.thenApply(ignored -> null);
+            }
+        }
+        return CompletableFuture.completedFuture(null);
```

Comment:
If you initialize lastSendFuture to CompletableFuture.completedFuture(null), you don't need the null check. Also, if batching is not enabled, you should still be waiting on lastSendFuture from the previous asyncSend (not sure if this is what you're referring to in your comment above).

Reply:
fixed